pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -328,485 +405,29 @@ func (gc *GarbageCollector) orphanFinalizer() {
-			continue
-		}
-		// TODO: we need to verify the reference resource is supported by the
-		// system. If it's not a valid resource, the garbage collector should i)
-		// ignore the reference when decide if the object should be deleted, and
-		// ii) should update the object to remove such references. This is to
-		// prevent objects having references to an old resource from being
-		// deleted during a cluster upgrade.
-		fqKind := schema.FromAPIVersionAndKind(reference.APIVersion, reference.Kind)
-		client, err := gc.clientPool.ClientForGroupVersionKind(fqKind)
-		if err != nil {
-			return err
-		}
-		resource, err := gc.apiResource(reference.APIVersion, reference.Kind, len(item.identity.Namespace) != 0)
-		if err != nil {
-			return err
-		}
-		owner, err := client.Resource(resource, item.identity.Namespace).Get(reference.Name)
-		if err == nil {
-			if owner.GetUID() != reference.UID {
-				glog.V(6).Infof("object %s's owner %s/%s, %s is not found, UID mismatch", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
-				gc.absentOwnerCache.Add(reference.UID)
-				continue
-			}
-			glog.V(6).Infof("object %s has at least an existing owner, will not garbage collect", item.identity.UID)
-			return nil
-		} else if errors.IsNotFound(err) {
-			gc.absentOwnerCache.Add(reference.UID)
-			glog.V(6).Infof("object %s's owner %s/%s, %s is not found", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
-		} else {
-			return err
-		}
-	}
-	glog.V(2).Infof("none of object %s's owners exist any more, will garbage collect it", item.identity)
-	return gc.deleteObject(item.identity)
-}
-
-func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
-	glog.Infof("Garbage Collector: Initializing")
-	for _, monitor := range gc.monitors {
-		go monitor.controller.Run(stopCh)
-	}
-
-	wait.PollInfinite(10*time.Second, func() (bool, error) {
-		for _, monitor := range gc.monitors {
-			if !monitor.controller.HasSynced() {
-				glog.Infof("Garbage Collector: Waiting for resource monitors to be synced...")
-				return false, nil
-			}
-		}
-		return true, nil
-	})
-	glog.Infof("Garbage Collector: All monitored resources synced. Proceeding to collect garbage")
-
-	// worker
-	go wait.Until(gc.propagator.processEvent, 0, stopCh)
-
-	for i := 0; i < workers; i++ {
-		go wait.Until(gc.worker, 0, stopCh)
-		go wait.Until(gc.orphanFinalizer, 0, stopCh)
-	}
-	Register()
-	<-stopCh
-	glog.Infof("Garbage Collector: Shutting down")
-	gc.dirtyQueue.ShutDown()
-	gc.orphanQueue.ShutDown()
-	gc.propagator.eventQueue.ShutDown()
+	return true
 }
 
 // *FOR TEST USE ONLY* It's not safe to call this function when the GC is still
 // busy.
-// GraphHasUID returns if the Propagator has a particular UID store in its
+// GraphHasUID returns if the GraphBuilder has a particular UID store in its
 // uidToNode graph. It's useful for debugging.
+// This method is used by integration tests.
```

Comment:
Actually this function is safe now. `uidToNode.Read()` holds a lock. Will update the doc.